### PR TITLE
Restructure hawc catalog in 2hwc and 3hwc

### PR DESCRIPTION
**Pull request by Jaleleddine**
Implementation of 2hwc and 3hwc. 2hwc works well. 3hwc needs bugs fixes.

### CODE REVIEW

File: gammapy/catalog/hawc.py [hunk 1]
```diff
@@ -196,7 +169,114 @@ class SourceCatalog2HWC(SourceCatalog):
...
+
+
+    def spectral_model(self, which="point"):
+        """Spectral model (`~gammapy.modeling.models.PowerLawSpectralModel`).
+
+        * ``which="point"`` -- Spectral model under the point source assumption.
+        * ``which="extended"`` -- Spectral model under the extended source assumption.
+          Only available for some sources. Raise ValueError if not available.
+        """
+        # index_errp = self.data["spec_{idx}_index_errp"]
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:225]: remove commented lines

File: gammapy/catalog/hawc.py [hunk 2]
```diff
@@ -196,7 +169,114 @@ class SourceCatalog2HWC(SourceCatalog):
...
+    The data is from tables 2 and 3 in the paper [1]_.
+
+    The catalog table contains 40 rows / sources.
+    The paper mentions 39 sources e.g. in the abstract.
+    The difference is due to Geminga, which was detected as two "sources" by the algorithm
+    used to make the catalog, but then in the discussion considered as one source.
+
+    References
+    ----------
+    .. [1] Abeysekara et al, "The 2HWC HAWC Observatory Gamma Ray Catalog",
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:269]: The reference will have to be updated.

**Review by registerrier — CHANGES_REQUESTED** [review]
Thanks @Jaleleddine . I have added some comments inline.

It is also necessary to add the new 3HWC classes to the the `CATALOG_REGISTRY` and `__all__` lists in `catalog/__init__.py` file.

Some specific tests for 3HWC will be needed too.

File: gammapy/catalog/hawc.py [hunk 3]
```diff
@@ -196,7 +169,114 @@ class SourceCatalog2HWC(SourceCatalog):
     description = "2HWC catalog from the HAWC observatory"
     """Catalog description"""
 
-    source_object_class = SourceCatalogObject2HWC
+    source_object_class = SourceCatalogObjectHWC
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:172]: Here you should keep `SourceCatalogObject2HWC` as this is the object used by this class.

File: gammapy/catalog/hawc.py [hunk 4]
```diff
@@ -196,7 +169,114 @@ class SourceCatalog2HWC(SourceCatalog):
...
+        }
+
+        model = Model.create("PowerLawSpectralModel", "spectral", **pars)
+
+        for name, value in errs.items():
+            model.parameters[name].error = value
+
+        return model
+
+class SourceCatalog3HWC():
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:255]: HEre you have to add the inheritance from `SourceCatalog`

File: gammapy/catalog/hawc.py [hunk 5]
```diff
@@ -6,10 +6,10 @@
 from gammapy.utils.scripts import make_path
 from .core import SourceCatalog, SourceCatalogObject
 
-__all__ = ["SourceCatalog2HWC", "SourceCatalogObject2HWC"]
+__all__ = ["SourceCatalog2HWC", "SourceCatalogObjectHWC"]
 
 
-class SourceCatalogObject2HWC(SourceCatalogObject):
+class SourceCatalogObjectHWC(SourceCatalogObject):
```
[the remark this answers is not in the document]
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py]: done

File: gammapy/catalog/hawc.py (hunk 1 above)
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py:226]: done

File: gammapy/catalog/hawc.py (hunk 2 above)
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py]: done

File: gammapy/catalog/hawc.py (hunk 4 above)
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py]: done

File: gammapy/catalog/hawc.py [hunk 6]
```diff
@@ -196,7 +169,114 @@ class SourceCatalog2HWC(SourceCatalog):
...
+        model = Model.create("PowerLawSpectralModel", "spectral", **pars)
+
+        for name, value in errs.items():
+            model.parameters[name].error = value
+
+        return model
+
+
+class SourceCatalogObject3HWC(SourceCatalogObjectHWC):
+
```
[the remark this answers is not in the document]
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py:217]: done

**Comment by Jaleleddine** [comment]
done

File: gammapy/catalog/__init__.py [hunk 7]
```diff
@@ -5,6 +5,8 @@
 from .fermi import *
 from .gammacat import *
 from .hawc import *
+from .hawc import SourceCatalogObject2HWC
```
Comment (registerrier) [inline gammapy/catalog/__init__.py:8]: Here you don' need to explicitly import the `SourceCatalogObject`. It must be properly defined in gammapy.catalog.hawc.py. See below

File: gammapy/catalog/__init__.py [hunk 8]
```diff
@@ -30,9 +34,11 @@
     "SourceCatalogObject3FGL",
     "SourceCatalogObject4FGL",
     "SourceCatalogObject2HWC",
+    "SourceCatalogObject3HWC",
     "SourceCatalogObjectGammaCat",
     "SourceCatalogObjectHGPSComponent",
     "SourceCatalogLargeScaleHGPS",
+    "SourceCatalogObject",
```
Comment (registerrier) [inline gammapy/catalog/__init__.py:41]: `SourceCatalogObject` is a generic class. It shouldn't be exposed. Remove

File: gammapy/catalog/hawc.py [hunk 9]
```diff
@@ -1,15 +1,16 @@
...
 """HAWC catalogs (https://www.hawc-observatory.org)."""
+import abc
 import numpy as np
 from astropy.table import Table
 from gammapy.modeling.models import Model, SkyModel
 from gammapy.utils.scripts import make_path
 from .core import SourceCatalog, SourceCatalogObject
 
-__all__ = ["SourceCatalog2HWC", "SourceCatalogObject2HWC"]
+__all__ = ["SourceCatalog2HWC", "SourceCatalog3HWC"]
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:10]: Here you also need to add `SourceCatalogObject2HWC` and `SourceCatalogObject3HWC` to expose them on the global API.

File: gammapy/catalog/hawc.py [hunk 10]
```diff
@@ -196,11 +170,118 @@ class SourceCatalog2HWC(SourceCatalog):
     description = "2HWC catalog from the HAWC observatory"
     """Catalog description"""
 
-    source_object_class = SourceCatalogObject2HWC
+    source_object_class = SourceCatalogObjectHWCBase
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:173]: No, here the object class is still `SourceCatalogObject2HWC`, not the Base class

File: gammapy/catalog/hawc.py [hunk 11]
```diff
@@ -196,11 +170,118 @@ class SourceCatalog2HWC(SourceCatalog):
...
+
+
+    def spectral_model(self, which="point"):
+        """Spectral model (`~gammapy.modeling.models.PowerLawSpectralModel`).
+
+        * ``which="point"`` -- Spectral model under the point source assumption.
+        * ``which="extended"`` -- Spectral model under the extended source assumption.
+          Only available for some sources. Raise ValueError if not available.
+        """
+        # index_errp = self.data["spec_{idx}_index_errp"]
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:226]: Comments are still there apparently.

File: gammapy/catalog/hawc.py [hunk 12]
```diff
@@ -196,11 +170,118 @@ class SourceCatalog2HWC(SourceCatalog):
...
+
+class SourceCatalog3HWC(SourceCatalog):
+    """HAWC 3HWC catalog.
+
+    One source is represented by `~gammapy.catalog.SourceCatalogObject3HWC`.
+
+    The data is from tables 2 and 3 in the paper [1]_.
+
+    The catalog table contains 65 rows / sources.
+    The difference is due to Geminga, which was detected as two "sources" by the algorithm
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:264]: Remove this sentence. It is out of scope for 3HWC

File: gammapy/catalog/hawc.py [hunk 13]
```diff
@@ -196,11 +170,118 @@ class SourceCatalog2HWC(SourceCatalog):
...
+       <https://data.hawc-observatory.org/datasets/3hwc-survey/index.php>`__
+    """
+
+    tag = "3hwc"
+    """Catalog name"""
+
+    description = "3HWC catalog from the HAWC observatory"
+    """Catalog description"""
+
+    source_object_class = SourceCatalogObjectHWCBase
```
Comment (registerrier) [inline gammapy/catalog/hawc.py:279]: Here `SourceCatalogObject3HWC` is the correct class.

**Review by registerrier — CHANGES_REQUESTED** [review]
Thanks @Jaleleddine 

There are still a few issues which are described with inline comments

File: gammapy/catalog/hawc.py (hunk 9 above)
Comment (Jaleleddine) [reply gammapy/catalog/hawc.py:10]: done

**Pull request merged**